### PR TITLE
EntityCompletenessTest

### PR DESCRIPTION
**Pull request by rjkip**

### CODE REVIEW

**Comment by rjkip** [comment]
@DRvanR

File: src/Surfnet/Conext/EntityVerificationFramework/Metadata/Contact.php [hunk 1]
```diff
@@ -96,19 +115,12 @@ public function __construct(
...
-            $valid = $valid && $this->email->equals($other->email);
-        }
-
-        return $valid && $this->givenName === $other->givenName && $this->surName === $other->surName;
+        return ($this->type === $other->type
+            || $this->type && $other->type && $this->type->equals($other->type))
+        && ($this->email === $other->email
+            || $this->email && $other->email && $this->email->equals($other->email))
+        && $this->givenName === $other->givenName
+        && $this->surName === $other->surName;
```
Comment (DRvanR) [inline src/Surfnet/Conext/EntityVerificationFramework/Metadata/Contact.php]: why not just use the `equals` check for `type` and `email` seeing they are now required objects?

File: src/Surfnet/Conext/EntityVerificationFramework/Metadata/Logo.php [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        return $logo;
+    }
+
+    private function __construct()
+    {
+    }
+
+    public function validate(Validator $validator, ValidationContext $context)
+    {
+        if (!$this->url->isValid()) {
```
Comment (DRvanR) [inline src/Surfnet/Conext/EntityVerificationFramework/Metadata/Logo.php]: Why not use `$this->url->validate($validator, $context)` here?

File: src/Surfnet/Conext/EntityVerificationFramework/Metadata/Logo.php [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            );
+        }
+    }
+
+    /**
+     * @return bool
+     */
+    private function isWidthValid()
+    {
+        return ((string) (int) $this->width) === $this->width && $this->width > 0;
```
Comment (DRvanR) [inline src/Surfnet/Conext/EntityVerificationFramework/Metadata/Logo.php]: `return ctype_digit($this->width) && $this->width > 0;` seems to do to same?

File: src/Surfnet/Conext/EntityVerificationFramework/Metadata/Logo.php [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            );
+        } elseif (!$this->url->matches('~^https://static\.surfconext\.nl/logos/idp/.+\.png$~')) {
+            $validator->addViolation(
+                sprintf(
+                    'Logo URL "%s" does not match https://static.surfconext.nl/logos/idp/<name>.png',
+                    $this->url
+                )
+            );
+        } else {
+            $response = $context->getHttpClient()->request('GET', $this->url->getValidUrl());
```
Comment (DRvanR) [inline src/Surfnet/Conext/EntityVerificationFramework/Metadata/Logo.php]: would it be timely to create something like `$context->getUrl($url)->isOk()` (or something like it e.g. `->statusCodeIs(Response::OK))? 

Also: `$context->postTo($url, $body[, $contentType='json'])->wasSuccesFul()`, etc. 
Just wondering, to keep it simple and all that.

File: src/Surfnet/Conext/EntityVerificationFramework/Metadata/Logo.php [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            if ($response->getStatusCode() !== 200) {
```
Comment (DRvanR) [inline src/Surfnet/Conext/EntityVerificationFramework/Metadata/Logo.php]: Would use the `Symfony\Component\HttpFoundation\Response::HTTP_OK` over `200` tbh

File: src/Surfnet/Conext/EntityVerificationFramework/Metadata/NameIdFormat.php [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+use Surfnet\Conext\EntityVerificationFramework\Assert;
+use Surfnet\Conext\EntityVerificationFramework\Metadata\Validator\ConfiguredMetadata\Validatable;
+use Surfnet\Conext\EntityVerificationFramework\Metadata\Validator\ConfiguredMetadata\ValidationContext;
+use Surfnet\Conext\EntityVerificationFramework\Metadata\Validator\ConfiguredMetadata\Validator;
+
+final class NameIdFormat implements Validatable
+{
+    const URN_SAML_11_UNSPECIFIED = 'urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified';
+    const URN_SAML_20_TRANSIENT = 'urn:oasis:names:tc:SAML:2.0:nameid-format:transient';
+    const URN_SAML_20_PERSISTENT = 'urn:oasis:names:tc:SAML:2.0:nameid-format:persistent';
```
Comment (DRvanR) [inline src/Surfnet/Conext/EntityVerificationFramework/Metadata/NameIdFormat.php]: Would it be an idea to include `simplesamlphp\saml2` for these kind of definitions? See [here][1]. Not sure, just wondering

[1]: https://github.com/simplesamlphp/saml2/blob/master/src/SAML2/Const.php#L130-L164

File: src/Surfnet/Conext/EntityVerificationFramework/Metadata/NameIdFormat.php [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    private function __construct()
+    {
+    }
+
+    public function validate(Validator $validator, ValidationContext $context)
+    {
+        if ($this->urn === null) {
+            $validator->addViolation('NameIDFormat is unknown');
```
Comment (DRvanR) [inline src/Surfnet/Conext/EntityVerificationFramework/Metadata/NameIdFormat.php]: Personnally would rephrase `unknown` as something like `not set`, `not configured`, `not given`... unknown vs unspecified, dunno, had to doublecheck they weren't the same.

File: src/Surfnet/Conext/EntityVerificationFramework/Metadata/Validator/ConfiguredMetadata/RootValidator.php [hunk 8]
```diff
@@ -16,42 +16,36 @@
  * limitations under the License.
  */
 
-namespace Surfnet\Conext\EntityVerificationFramework\Value;
+namespace Surfnet\Conext\EntityVerificationFramework\Metadata\Validator\ConfiguredMetadata;
 
 use Surfnet\Conext\EntityVerificationFramework\Assert;
 
-final class ImageList
+final class RootValidator implements Validator
```
Comment (DRvanR) [inline src/Surfnet/Conext/EntityVerificationFramework/Metadata/Validator/ConfiguredMetadata/RootValidator.php]: `final class Validator implements ConfiguredMetadataValidator`?

File: src/Surfnet/Conext/EntityVerificationFramework/Runner.php [hunk 9]
```diff
@@ -143,9 +145,11 @@ public function run(VerificationReporter $reporter, VerificationSuiteWhitelist $
         }
 
         $this->logger->info(sprintf(
-            'Completed Run of Entity Verification Framework, "%d" Entities Verified with "%d" suites.',
+            'Completed Run of Entity Verification Framework, ' .
+            'verification of "%d" Entities with "%d" suites resulted in "%d" reports',
```
Comment (DRvanR) [inline src/Surfnet/Conext/EntityVerificationFramework/Runner.php]: nice addition!

File: src/Surfnet/Conext/EntityVerificationFramework/Tests/Metadata/NameIdFormatTest.php [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        NameIdFormat::fromUrn($nonString);
+    }
+
+    /**
+     * @test
+     * @group value
+     */
+    public function it_can_equal_other_formats()
+    {
+        $format0 = NameIdFormat::fromUrn('urn:mad:bro');
```
Comment (DRvanR) [inline src/Surfnet/Conext/EntityVerificationFramework/Tests/Metadata/NameIdFormatTest.php]: lol

File: src/Surfnet/Conext/EntityVerificationFramework/Tests/Metadata/NameTest.php [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+use Surfnet\Conext\EntityVerificationFramework\Metadata\Validator\ConfiguredMetadata\ValidationContext;
+use Surfnet\Conext\EntityVerificationFramework\Metadata\Validator\ConfiguredMetadata\Validator;
+
+class NameTest extends TestCase
+{
+    /**
+     * @test
+     * @group value
+     */
+    public function it_reports_violations_when_locales_en_andor_nl_are_not_filled()
```
Comment (DRvanR) [inline src/Surfnet/Conext/EntityVerificationFramework/Tests/Metadata/NameTest.php]: This only tests that if both are absent, two violations are reported :stuck_out_tongue_winking_eye:

File: src/Surfnet/Conext/EntityVerificationFramework/Tests/Metadata/PublishedMetadataListTest.php [hunk 12]
```diff
@@ -66,7 +65,7 @@ public function it_can_find_by_entity()
         $metadataXmlString = file_get_contents(__DIR__ . '/fixtures/metadata-multi-engineblock-onegini.xml');
         $metadataXml = simplexml_load_string($metadataXmlString);
 
-        $onegini = new Entity(new EntityId('https://www.onegini.me'), EntityType::IdP());
+        $onegini = new \Surfnet\Conext\EntityVerificationFramework\Value\Entity(new EntityId('https://www.onegini.me'), EntityType::IdP());
```
Comment (DRvanR) [inline src/Surfnet/Conext/EntityVerificationFramework/Tests/Metadata/PublishedMetadataListTest.php]: FQCN?

File: src/Surfnet/Conext/EntityVerificationFramework/Value/Entity.php [hunk 13]
```diff
@@ -52,11 +52,38 @@ public static function fromDescriptor(array $descriptor)
...
+    public function hasEntityId(EntityId $entityId)
+    {
+        return $this->entityId->equals($entityId);
+    }
+
+    // @codingStandardsIgnoreStart
+    /**
+     * @return bool
+     */
+    public function isSP()
```
Comment (DRvanR) [inline src/Surfnet/Conext/EntityVerificationFramework/Value/Entity.php]: seeing how often we do this (`@codingStandardIgnore`), thoughts about `isServiceProvider` & `isIdentityProvider`?

**Comment by DRvanR** [comment]
Some minor things, but overall great work!

File: src/Surfnet/Conext/EntityVerificationFramework/Metadata/NameIdFormat.php (hunk 6 above)
Comment (rjkip) [inline src/Surfnet/Conext/EntityVerificationFramework/Metadata/NameIdFormat.php]: Nah. They're constants from a spec.

File: src/Surfnet/Conext/EntityVerificationFramework/Metadata/NameIdFormat.php (hunk 7 above)
Comment (rjkip) [inline src/Surfnet/Conext/EntityVerificationFramework/Metadata/NameIdFormat.php]: As in in the violation message?

Comment (DRvanR) [inline src/Surfnet/Conext/EntityVerificationFramework/Metadata/NameIdFormat.php]: yesh

File: src/Surfnet/Conext/EntityVerificationFramework/Metadata/Logo.php (hunk 4 above)
Comment (rjkip) [inline src/Surfnet/Conext/EntityVerificationFramework/Metadata/Logo.php]: Let's see how Guzzle will be used in other places (also in regard to more detailed violation messages than "it no work") before creating these convenience methods. No rush.

**Comment by rjkip** [comment]
@arothuis

File: src/Surfnet/Conext/EntityVerificationFramework/Tests/Metadata/ContactTypeValidationTest.php [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+        $contactType = ContactType::fromString(ContactType::TYPE_SUPPORT);
+        $contactType->validate($validator, $context);
+    }
+
+    /**
+     * @param string $expectedToContain
+     * @return ClosureMatcher
+     */
+    private static function containsMatcher($expectedToContain)
```
Comment (arothuis) [inline src/Surfnet/Conext/EntityVerificationFramework/Tests/Metadata/ContactTypeValidationTest.php]: I understand where you're coming from, as a matcher is returned that allows Mockery to check whether a parameter `contains` a certain string, but `containsMatcher` seems a like an assertion of some kind.

File: src/Surfnet/Conext/EntityVerificationFramework/Tests/Metadata/LogoValidationTest.php [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+{
+    /**
+     * @test
+     * @group value
+     * @dataProvider logosWithViolations
+     *
+     * @param Logo   $logo
+     * @param string $violation
+     */
+    public function it_can_be_invalid(Logo $logo, $violation)
```
Comment (arothuis) [inline src/Surfnet/Conext/EntityVerificationFramework/Tests/Metadata/LogoValidationTest.php]: Not sure about the naming here. What does `it` point to and what does it mean when it can be invalid? `logo_can_be_validated`?

**Comment by arothuis** [comment]
A few things (as discussed), other than that :+1:

**Comment by rjkip** [comment]
@DRvanR @arothuis

**Comment by arothuis** [comment]
:+1:

**Comment by DRvanR** [comment]
:+1:  as is, up to your discretion whether or not to make logoUrl etc.
